wgpu-hal/src/dx12/command.rs

```diff
@@ -380,6 +379,101 @@ impl crate::CommandEncoder<super::Api> for super::CommandEncoder {
         }
     }
 
+    unsafe fn clear_texture(
+        &mut self,
+        texture: &super::Texture,
+        subresource_range: &wgt::ImageSubresourceRange,
+    ) {
+        // Note that CopyTextureRegion for depth/stencil or multisample resources would require full subresource copies.
+        // Meaning we'd need a much larger pre-zeroed buffer
+        // (but instead we just define clear_texture to not support these)
+
+        let list = self.list.unwrap();
+        let mut src_location = d3d12::D3D12_TEXTURE_COPY_LOCATION {
+            pResource: self.shared.zero_buffer.as_mut_ptr(),
+            Type: d3d12::D3D12_TEXTURE_COPY_TYPE_PLACED_FOOTPRINT,
+            u: mem::zeroed(),
+        };
+        let mut dst_location = d3d12::D3D12_TEXTURE_COPY_LOCATION {
+            pResource: texture.resource.as_mut_ptr(),
+            Type: d3d12::D3D12_TEXTURE_COPY_TYPE_SUBRESOURCE_INDEX,
+            u: mem::zeroed(),
+        };
+        let raw_format = conv::map_texture_format(texture.format);
+        let format_desc = texture.format.describe();
+
+        let depth = if texture.dimension == wgt::TextureDimension::D3 {
+            texture.size.depth_or_array_layers
+        } else {
+            1
+        };
+
+        let mip_range = subresource_range.base_mip_level
+            ..(subresource_range.base_mip_level
+                + subresource_range
+                    .mip_level_count
+                    .map_or(texture.mip_level_count, |c| c.get()));
+        let array_range = subresource_range.base_array_layer
+            ..(subresource_range.base_array_layer
+                + subresource_range
+                    .array_layer_count
+                    .map_or(texture.array_layer_count(), |c| c.get()));
+
+        for mip_level in mip_range {
+            let bytes_per_row = texture.mip_level_size(mip_level).width
+                / format_desc.block_dimensions.0 as u32
+                * format_desc.block_size as u32;
+            // round up to a multiple of d3d12::D3D12_TEXTURE_DATA_PITCH_ALIGNMENT
+            let bytes_per_row = (bytes_per_row + d3d12::D3D12_TEXTURE_DATA_PITCH_ALIGNMENT - 1)
+                / d3d12::D3D12_TEXTURE_DATA_PITCH_ALIGNMENT
+                * d3d12::D3D12_TEXTURE_DATA_PITCH_ALIGNMENT;
+
+            let max_rows_per_copy = super::ZERO_BUFFER_SIZE as u32 / bytes_per_row;
+            // round down to a multiple of rows needed by the texture format
+            let max_rows_per_copy = max_rows_per_copy / format_desc.block_dimensions.1 as u32
+                * format_desc.block_dimensions.1 as u32;
+            assert!(max_rows_per_copy > 0, "Zero buffer size is too small to fill a single row of a texture with dimension {:?}, size {:?} and format {:?}", texture.dimension, texture.size, texture.format);
+
+            for array_layer in array_range.clone() {
```

Comment:
is `clone` needed here?

Reply:
yeah oddly enough it is. Compiler complains that this is an implicit into_iter() call and as such it is moved from the previous loop run. And range doesn't have an iter() call as alternative